### PR TITLE
fix(security): derive control-char stripping from Unicode's format category

### PR DESCRIPTION
**Pull request by WingedGuardian**
## What

`strip_control_chars` normalizes free-text scalars — campaign names and awareness-signal `name`/`source`/`baseline_note` — before they are rendered one-per-line into a line-parsed prompt, so injected text cannot forge or conceal a prompt line.

Its character class was **hand-enumerated and covered 13 of Unicode's 170 `Cf` codepoints**, silently omitting concealment characters from the very families it did cover. The clearest example is **U+061C ARABIC LETTER MARK** — a bidi mark whose siblings LRM/RLM were already stripped, so the Trojan-source defence was incomplete on its own axis. Also missing: SOFT HYPHEN, WORD JOINER, the invisible-operator block, interlinear annotation, and the U+E0000 tag block.

## Change

The set is now **derived from Python's Unicode database** rather than hand-picked. Two tests keep it honest:
- a **drift test** that regenerates the set from `unicodedata` and fails if the two diverge, so a future UCD bump cannot silently reopen the gap;
- a **monotonicity test** proving every codepoint the old class stripped is still stripped.

### The two characters deliberately kept

U+200C ZWNJ and U+200D ZWJ are `Cf` but are **not** stripped: doing so would break every emoji ZWJ sequence (👨‍👩‍👧 → three separate people) and corrupt Persian and Indic words, where ZWNJ is orthographically required.

That exclusion is safe for this module's threat model: neither is a `str.splitlines()` boundary, so a kept joiner cannot forge a prompt line, and unlike the bidi overrides (still stripped) it does not conceal adjacent printable text. **Nothing in the suite covered either joiner before**, so a naive whole-category rewrite would have shipped green while mangling real content — those preservation tests are now in place.

## Campaign name heal

Names have been sanitized at the write boundary since the previous release, so nothing new lands malformed; rows written earlier were never repaired.

This runs inside campaign init, **before `runner.start()`** — the ordering is the whole point. The scheduler derives each job id from the stored name, and the reverted `d0012` migration renamed rows at data-migration time, well after init had already registered those jobs. Collisions on the UNIQUE `name` column, names that normalize to empty, and repeat runs all leave the row untouched rather than merging or blanking it.

## Verification

- 56 targeted tests green (sanitizer, name heal, campaign CRUD, awareness signal format); `ruff` clean.
- **Verify-RED** on every new test, including deliberately removing the empty-name guard and the ordering guard to confirm each is caught.
- E2E through the real `SignalReading` object: ordinary names untouched, ALM and soft hyphen now stripped, ZWJ emoji and Persian ZWNJ preserved, newline still collapsed.
- Adversarial security review executed the character-set check against all 0x110000 codepoints (strict superset, **0 regressions**), checked every fragment boundary in the concatenated class, and verified the init ordering *functionally* — `runner.start()` issues a fresh campaign query after the heal commits.

## Deploy path

Runtime code: `git pull` + server restart. The heal runs at the next boot; it is idempotent and a no-op where nothing is dirty.

## Summary by CodeRabbit

- **Bug Fixes**
  - Cleaned legacy campaign names containing invisible control characters during startup.
  - Preserved visible script characters, emoji, and joiners while removing hidden formatting controls.
  - Rejected names that become empty or conflict with reserved scheduler names.
  - Handled name collisions safely without overwriting existing campaigns.
  - Kept related health and event history aligned when names are repaired.
- **Security**
  - Improved protection against hidden Unicode characters and forged log lines.
- **Reliability**
  - Ensured cleanup failures do not prevent campaign startup.

### CODE REVIEW

**Comment by WingedGuardian** [comment]
@codex review

**Comment by WingedGuardian** [comment]
## Automated first-pass review

**Domain 1 — Structural review**

Checked all changed `.py` files (`src/genesis/campaigns/name_heal.py`, `src/genesis/runtime/init/campaigns.py`, `src/genesis/security/sanitizer.py`, and the two test files) against the pattern/import/coverage/documentation/convention checklist.

- No ERROR-level pattern violations (no `nohup`, no raw SQL against production tables outside CRUD, no bare `except:`, no mutable defaults, no `git add -A`/force-push patterns, etc.). `update_campaign`/`list_campaigns` are used via the CRUD layer as intended; the raw `db.execute(INSERT ...)` in `tests/test_campaigns/test_name_heal.py` is confined to a test fixture and is there specifically to bypass the write-boundary strip (documented in the file's docstring) — this is the standard/expected way to reproduce a pre-fix row, not a production code path.
- Import hygiene clean — lazy imports inside the function bodies (`name_heal.py`, `runtime/init/campaigns.py`) follow the existing lazy-import convention in this init module, no circular imports, no star imports.
- Coverage: new `name_heal.py` has a matching `tests/test_campaigns/test_name_heal.py`; the changed `strip_control_chars` regex has matching updated/added tests in `tests/test_security/test_strip_control_chars.py`.
- Documentation: `heal_campaign_names` has a full docstring; the `_CF_STRIPPED`/`_CONTROL_RUN_RE` change is documented inline; `strip_control_chars`'s docstring still accurately describes behavior.
- Ordering claim verified by reading `src/genesis/runtime/init/campaigns.py` directly: `await heal_campaign_names(rt._db)` (line 40) executes before `runner = CampaignRunner(...)` / `await runner.start()` (lines 46–52), and the heal call is wrapped in its own fail-open `try/except` so a heal failure can't block campaign init.
- Independently re-derived the security fix rather than trusting the PR's own claim: ran `_CONTROL_RUN_RE` against every codepoint 0x0–0x10FFFF via `unicodedata.category()` (Python's UCD 14.0.0) — confirmed the pattern strips the entire `Cf` category except `U+200C`/`U+200D`, with zero missing and zero wrongly-stripped codepoints, matching what `test_cf_ranges_match_unicodedata` asserts.

Structural review: **CLEAN** — no pattern violations found.

**Domain 2 — PII/Secrets/Wording**

- No API keys, tokens, credentials, private IPs (10.x/192.168.x/100.64–127.x), IPv6 ULA, SSH keys, or DB connection strings in the added lines.
- No personal emails, usernamed paths, location-revealing timezones, cloud account IDs, hardware identifiers, or private hostnames. (`/tmp/s.md` in the test fixture is a fake placeholder path, not a real temp-file usage.)
- PR title/body/CHANGELOG: describes a defense-in-depth improvement to an existing, already-public sanitizer function (expanding hand-picked Unicode coverage to the full `Cf` category) and a data-repair heal for legacy rows. No leaked credentials, no private infrastructure, no specific incident referenced, no unpatched-bypass detail — the gap being described is closed by this same diff.

PII/Secrets/Wording scan: **CLEAN**



---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by WingedGuardian** [comment]
@codex review

All three verified against the code and the UCD, all three correct.

**Rendered Cf marks** — the deepest one. While verifying I measured that **no `Cf` codepoint is a `str.splitlines()` boundary** (checked all 0x110000), so line forging is closed entirely by the C0/C1 + U+2028/U+2029 ranges; the Cf set only closes concealment and reordering. Keeping the rendered marks therefore costs nothing on the line-forging axis. The rule is now **invisibility** rather than category, still derived from `unicodedata`: strip iff bidi class is BN, or an explicit bidi override/isolate, or an invisible direction mark (LRM/RLM/ALM — strong class but zero-width, so class alone misses them), or an interlinear annotation control; minus ZWJ/ZWNJ. Monotonicity re-proven: every codepoint the pre-PR regex stripped is still stripped.

**Empty-after-strip** — rejected at the storage chokepoint (`create_campaign` and `update_campaign`), so no caller can write one; the MCP entry returns a clean error.

**job_health orphan** — the row now follows the rename, best-effort and non-fatal, and refuses to clobber existing history under the new name.

75 targeted tests green; new locks for rendered-marks-survive, no-Cf-forges-a-line, job_health-follows-rename, job_health-collision, and empty-name rejection.

**Comment by WingedGuardian** [comment]
@coderabbitai review

**Comment by WingedGuardian** [comment]
@coderabbitai review

File: src/genesis/security/sanitizer.py [hunk 1]
```diff
@@ -41,12 +41,34 @@ def strip_boundary_markers(text: str) -> str:
...
+# every emoji ZWJ sequence (stripping it turns 👨‍👩‍👧 into three separate people)
+# and ZWNJ is orthographically required in Persian and Indic scripts (stripping it
+# changes the WORD, not just its rendering). Removal has to be context-aware; a
+# whole-category strip corrupts legitimate content.
+_CF_STRIPPED = (
+    r"\u00ad\u0600-\u0605\u061c\u06dd\u070f\u0890-\u0891\u08e2"
+    r"\u180e\u200b\u200e-\u200f\u202a-\u202e\u2060-\u2064"
+    r"\u2066-\u206f\ufeff\ufff9-\ufffb\U000110bd\U000110cd"
+    r"\U00013430-\U0001343f\U0001bca0-\U0001bca3"
+    r"\U0001d173-\U0001d17a\U000e0001\U000e0020-\U000e007f"
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/security/sanitizer.py]: Already addressed at head — this thread was never answered, which is why it still reads as open.

`_CF_INVISIBLE` (`sanitizer.py:73-78`) is a curated subset, not all of Cf. Every character this finding names is deliberately absent from it: the Arabic number/footnote/end-of-ayah marks (U+0600-0605, U+06DD, U+08E2, U+0890-0891), SYRIAC ABBREVIATION MARK, the Kaithi number signs, and the Egyptian hieroglyph joiners. `sanitizer.py:61-68` states the rule in place — strip a Cf codepoint only if it is *invisible* (bidi class BN, an explicit override/isolate, an invisible direction mark, or an interlinear annotation control); everything else in Cf is rendered script content and is kept.

ZWNJ and ZWJ are the two deliberate exceptions in the other direction: both are BN, but ZWJ builds every emoji ZWJ sequence and ZWNJ is orthographically required in Persian and Indic scripts, so both are preserved. `test_emoji_name_survives_the_heal` pins that.

File: src/genesis/security/sanitizer.py [hunk 2]
```diff
@@ -41,12 +41,34 @@ def strip_boundary_markers(text: str) -> str:
...
+)
+
 _CONTROL_RUN_RE = re.compile(
     "["
     r"\x00-\x1f\x7f-\x9f"  # C0 + C1 control chars (incl. tab/newline/CR, NEL)
     r"\u2028\u2029"  # LINE / PARAGRAPH SEPARATOR (str.splitlines boundaries)
-    r"\u200b\u200e\u200f\u202a-\u202e\u2066-\u2069\ufeff"  # zero-width / bidi format
-    "]+"
+    + _CF_STRIPPED  # every Cf format control EXCEPT ZWJ/ZWNJ (see above)
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/security/sanitizer.py]: Already addressed at head — unanswered thread rather than an open defect.

`campaign_tools.py:87` guards `if not name:` immediately after the strip and returns an error rather than inserting, so the unaddressable-campaign path the finding describes cannot be reached. The CRUD boundary raises as well, pinned by `test_empty_after_strip_is_rejected_at_the_crud_boundary`.

Checked the sibling path this finding did not name, since the class is "a name reaching the DB unvalidated": `_impl_campaign_update` (`campaign_tools.py:317`) takes `name` as a *lookup* key and never writes a new one, so it carries no equivalent hazard.

File: src/genesis/campaigns/name_heal.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            continue
+        if clean in taken:
+            logger.warning(
+                "Campaign %s: normalized name collides with an existing campaign "
+                "— left as-is (rename it manually to resolve)",
+                row.get("id"),
+            )
+            continue
+        try:
+            await crud.update_campaign(db, row["id"], name=clean)
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/campaigns/name_heal.py:85]: Already addressed at head — `_migrate_job_health` (`name_heal.py:81`) carries the row across the rename and is called from the heal loop, so the fossil-under-the-old-name case this finding describes no longer occurs on the ordinary path. Pinned by `test_job_health_row_follows_the_rename`, which asserts both that the history arrives under the new name and that nothing is left under the old one.

Your two follow-up findings on that same function — the collision branch and the run-event history — were both real and are fixed in `8191b960`; see those threads.

File: src/genesis/campaigns/name_heal.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        cur = await db.execute(
+            "SELECT 1 FROM job_health WHERE job_name = ?", (new_job,)
+        )
+        if await cur.fetchone():
+            logger.warning(
+                "job_health row already exists for %s — leaving %s in place",
+                new_job,
+                old_job,
+            )
+            return
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/campaigns/name_heal.py]: Confirmed and fixed in `8191b960` — but **I did the opposite of the suggested remedy**, so the reasoning is worth stating.

The finding says "Preserve the new row if it must win". At heal time nothing can legitimately hold `campaign_{new_name}`:

- the campaign answered to the OLD name until a moment ago;
- `campaigns.name` is UNIQUE, so no other campaign holds the new one;
- the heal runs BEFORE `runner.start()` — asserted structurally by the pre-existing `test_heal_runs_before_runner_start` — so nothing has written health under the new key in this process.

A row sitting there can therefore only be a fossil of a campaign that previously held the name and was deleted. The LIVE history is the one under the old name. So the old behaviour preserved the record that can never be written to again and orphaned the one that will be. It is now the reverse: the fossil is deleted, the live row moves in, and the discard is logged with its reason.

Your point about *why* it matters is exactly right and is what makes this worth doing rather than tolerating — `get_stale_jobs` filters on neither the live scheduler registry nor recency, so an orphan emits `job_stale` on every sweep forever and enters the ego reconciliation snapshot, with nothing to purge it.

**Flagging for your next pass:** `test_job_health_collision_leaves_both_rows` asserted the old behaviour (`total_runs == 99`, the fossil winning). It is rewritten rather than deleted, as `test_job_health_collision_keeps_the_live_history`, with the inversion argued in its docstring. A rewritten assertion is the shape that hides a regression, so it deserves more scrutiny than the new tests.

Verify-RED: mutating the `DELETE` back to a bare `return` — i.e. restoring exactly the old behaviour — turns that test RED.

File: src/genesis/campaigns/name_heal.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        await db.execute(
+            "UPDATE job_health SET job_name = ? WHERE job_name = ?",
+            (new_job, old_job),
+        )
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/campaigns/name_heal.py:195]: Confirmed and fixed in `8191b960`. `job_run_events` now moves in the same transaction as `job_health`.

Checked the constraint before writing the bulk UPDATE rather than assuming: `job_run_events.job_name` is plain `TEXT NOT NULL` with the PK on `id` (`db/schema/_tables.py:2000-2007`) and only a non-unique index on `(job_name, recorded_at)`, so re-pointing cannot collide — which also means any events already under the new name are re-pointed alongside the live series rather than dropped. That direction is deliberate: an over-complete run series still grades, a truncated one voids.

Verify-RED: appending `AND 0` to the new UPDATE turns `test_job_run_events_follow_the_rename` RED.

One correction to my own first attempt, since it bears on reading the test: it initially expected `campaign_nightlysweep` and failed. That was my error, not the fix — `strip_control_chars` SUBSTITUTES a space for a control run (`sanitizer.py:107`), it does not delete, so the healed name is `nightly sweep`.

File: src/genesis/campaigns/name_heal.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                "Campaign %s: name normalizes to empty — left as-is", row.get("id")
+            )
+            continue
+        if clean in taken:
+            logger.warning(
+                "Campaign %s: normalized name collides with an existing campaign "
+                "— left as-is (rename it manually to resolve)",
+                row.get("id"),
+            )
+            continue
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/campaigns/name_heal.py:67]: Confirmed and fixed in `8191b960` — and the finding is **narrower than the bug**, so it is fixed as a class rather than at the site you named.

Reading `runner.py:66-82`: `start()` schedules every campaign as `campaign_{name}` and THEN registers the reaper as `campaign_pending_reaper`, both with `replace_existing=True`. Registration order decides the winner, so the reaper evicts the campaign — silently, which is the worst part: no exception, no log, the campaign simply never ticks again. `replace_existing` is the behaviour we want everywhere else (it makes re-registration idempotent), which is precisely why nothing complains.

But the heal is not the only door. `campaign_create` would accept the name directly and produce the identical dead campaign, with no healing involved. Fixing only the heal would have left that open.

So:

- `RESERVED_CAMPAIGN_NAMES` is exported from `runner.py` — the module that owns the reservation — so the enforcement points cannot drift from the registration. A reserved name only one of them knows about is how this reopens.
- **Write boundary** (`campaign_tools.py`): rejected with an error that says why. This is the only place a human is told.
- **Heal** (`name_heal.py`): a stored name that would normalize into a reserved value is left dirty and logged. That direction is deliberate — a dirty name still runs, so leaving it dirty strictly beats healing it into a name that gets evicted.

The check runs AFTER the strip, so a decorated variant (`pending_reaper` + ZWSP) cannot slip past and normalize into the reserved value later; the new test exercises both forms.

Verify-RED: disabling either guard turns its test RED (`if clean in RESERVED_CAMPAIGN_NAMES:` → `if False:`, and the same at the write boundary).

**Comment by WingedGuardian** [comment]
@codex review

File: src/genesis/campaigns/name_heal.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+                row.get("id"),
+                clean,
+            )
+            continue
+        try:
+            await crud.update_campaign(db, row["id"], name=clean)
+        except Exception:
+            logger.exception("Campaign %s: name heal failed", row.get("id"))
+            continue
+        await _migrate_job_health(db, name, clean)
```
[the remark this answers is not in the document]
Comment (WingedGuardian) [reply src/genesis/campaigns/name_heal.py]: Confirmed and fixed in `7f1f352f`. You are right that it is permanent rather than transient, and that is the sharp part: `crud.update_campaign` commits (`db/crud/campaigns.py:123`), so by the time the migration runs the rename is durable. Next startup `clean == name`, the row is skipped before reaching the migration at all, and the old keys stay orphaned forever — the exact state this module exists to remove, reintroduced through its own error path.

`_migrate_job_health` now returns a bool instead of swallowing, and the caller undoes the rename on failure, restoring the pair to "both un-migrated" so the row is eligible for a retry. If the revert *also* fails, that is logged explicitly as an orphan that will not self-heal rather than passing silently.

**One thing worth recording, because reading the code will not show it.** The first revert went through `crud.update_campaign` and the test failed with `'retry me' == 'retry\tme'`. The CRUD write boundary *sanitizes the name* — so handing it the original dirty name writes the clean one straight back, and the revert is a silent no-op. The row stays renamed and still never retries.

Restoring a pre-boundary value is the one case that legitimately has to go around the boundary, so the revert uses a raw UPDATE, with the reason stated at the call site. It is the same reason this module's tests insert their fixtures with raw SQL. A test asserting only "heal returned 0" would have passed against a revert that did nothing.

**Verify-RED, including a mutation that was too coarse to count.** The first mutation put `if False:` on `if not await _migrate_job_health(...)`, which gave 4 failures rather than 1 — the call sits inside the condition, so it disabled the migration as well as the revert. A mutation that breaks more than the thing under test proves less, not more: it cannot distinguish "the revert is pinned" from "the migration is pinned".

Redone precisely — route the revert's UPDATE through `strip_control_chars`, i.e. exactly the naive version that looks right and does nothing:

```
1 failed, 14 passed  — only test_failed_state_migration_reverts_the_rename_so_it_retries
```

So the test pins the boundary bypass specifically. Restored from a byte copy afterwards and checked for residue rather than assumed: 0 occurrences of the mutation markers, ruff clean, 51 passed across the three affected files.
